src/context.cpp

```diff
@@ -194,25 +195,43 @@ const memory::unordered_map<rule *, context::object_set> &context::filter_inputs
         input_filter::cache_type &cache = it->second;
         auto exclusion = filter->match(store_, cache, deadline);
         if (exclusion.has_value()) {
+            DDWAF_DEBUG("Has value");
             for (const auto &rule : exclusion->rules) {
-                auto exclude_it = rules_to_exclude.find(rule);
-                if (exclude_it != rules_to_exclude.end() &&
-                    exclude_it->second == filter_mode::bypass) {
-                    continue;
+                if (!exclusion->objects.persistent.empty()) {
+                    auto &rule_policy = exclusion_policy_.persistent[rule];
+                    if (rule_policy.mode == filter_mode::bypass) {
+                        // If the rule has been bypassed, there is no need to
+                        // add persistent or ephemeral objects to it.
+                        DDWAF_DEBUG("Skipping rule");
+                        continue;
+                    }
+                    rule_policy.objects.insert(
+                        exclusion->objects.persistent.begin(), exclusion->objects.persistent.end());
+                } else {
+                    auto it = exclusion_policy_.persistent.find(rule);
+                    if (it != exclusion_policy_.persistent.end() &&
+                        it->second.mode == filter_mode::bypass) {
+                        continue;
+                    }
                 }
 
-                auto &common_exclusion = objects_to_exclude_[rule];
-                common_exclusion.insert(exclusion->objects.begin(), exclusion->objects.end());
+                if (!exclusion->objects.ephemeral.empty()) {
+                    auto &rule_policy = exclusion_policy_.ephemeral[rule];
+                    // Bypass has precedence over monitor
+                    if (rule_policy.mode != filter_mode::bypass) {
+                        rule_policy.objects.insert(exclusion->objects.ephemeral.begin(),
+                            exclusion->objects.ephemeral.end());
+                    }
+                }
```

Comment:
again, it feels like moving the logic to something like `exclusion_policy_.add_input_exclusion(rule, object_set)` would help

Reply:
I agree, the main reason I didn't do it is because I want to have an abstraction on top of exclusion filters which should also make it simpler to test things. However, I don't see any harm in having this function in the `exclusion::context_policy`, so I'll add it.